OpenRA.Game/Primitives/SegmentStream.cs

```
@@ -59,7 +59,15 @@ public override long Position
 			set { BaseStream.Position = BaseOffset + value; }
 		}
 
-		public override int Read(byte[] buffer, int offset, int count) { return BaseStream.Read(buffer, offset, count); }
+		public override int Read(byte[] buffer, int offset, int count)
+		{
+			// ensure caller can't read past of the end of our stream
+			var limit = BaseCount - Position + offset - count;
+			if (limit < 0) count += (int)limit;
```

It is standard behaviour that attempting to read past the end of a stream will throw an exception, so we should do that here too instead of silently clamping it.

---

I'm not sure I agree with it being standard behaviour

https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?redirectedfrom=MSDN&view=netframework-4.7.1#System_IO_Stream_Read_System_Byte___System_Int32_System_Int32_

To me this should return zero if there is nothing left in the buffer and no more is expected.